stubs/psycopg2/psycopg2/__init__.pyi

```diff
@@ -32,6 +33,18 @@ from psycopg2._psycopg import (
     threadsafety as threadsafety,
 )
 
+_T_conn = TypeVar("_T_conn", bound=connection)
+
+@overload
+def connect(dsn: str, connection_factory: Callable[..., _T_conn], cursor_factory: None = ..., **kwargs: Any) -> _T_conn: ...
+@overload
+def connect(
+    dsn: str | None = ..., *, connection_factory: Callable[..., _T_conn], cursor_factory: None = ..., **kwargs: Any
+) -> _T_conn: ...
+@overload
 def connect(
-    dsn: Any | None = ..., connection_factory: Any | None = ..., cursor_factory: Any | None = ..., **kwargs
+    dsn: str | None = ...,
+    connection_factory: Callable[..., connection] | None = ...,
+    cursor_factory: Callable[..., cursor] | None = ...,
+    **kwargs: Any,
 ) -> connection: ...
```

Comment:
I experimented a bit with giving `connection` a generic type param. For example:

```python
_T_conn_cur = TypeVar("_T_conn_cur", bound=cursor)

class connection(Generic[_T_conn_cur]):
    # ...

    @overload
    def cursor(self, name: str | bytes | None = ..., *, withhold: bool = ..., scrollable: bool | None = ...) -> _T_conn_cur: ...

    # ...
```

That way, in this overload of `connect`, the `cursor_factory` param could have a type of `Callable[..., _T_cur]` and its return type could be changed to `connection[_T_cur]`, but this introduces a few problems. The lack of support for a default value in `TypeVar`s (i.e. so that `connection` and `connection[cursor]` would be equivalent) is a nuisance, but the most serious problem is that the other two `connect` overloads would end up accepting, for example, `connection_factory=connection[NamedTupleCursor]`, which would produce a calculated return type that is inaccurate since the underlying code does not recognize the type param.